File: detect_baseball.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <vector>

using namespace std;
using namespace cv;

void detect_ball(
   const Mat& frame1,
   const Mat& frame2,
   Mat& ball_frame,
   float& r,
   Point& cent) {

   
   Mat diff = frame2-frame1;
   int numBallPix = sum(diff > 100)[0]; 
   if (numBallPix > 150) {
      // We have a ball or something here
      Mat bg = diff < 30; //30
      //imshow("frame2",frame2);
      //imshow("test",bg);
      //waitKey(30);
      Mat strel = getStructuringElement(MORPH_RECT,Size(25,25));
      Mat bg_erode; 
      erode(bg,bg_erode,strel);
      Mat frame2_tmp;
      frame2.copyTo(frame2_tmp,bg_erode==0);
       
      // Let's get the ball only
      Mat nonbg_tmp = (bg_erode==0);
      Mat mask_nz = (frame2>0);
      Mat nonbg = nonbg_tmp.mul(mask_nz);
      //imshow("nonbg",nonbg);
      //imshow("mask_nz",mask_nz);
      //waitKey(0);
      Mat labels;
      Mat stats;
      Mat centroids;
      int numComponents = connectedComponentsWithStats(nonbg,labels,stats,
         centroids,8);

      // Get second and third largest index
      // Clearly not the most elegant way to do this
      int first = 0;
      int second = 0;
      int third = 0;
      int ii1,ii2,ii3;
      int tmp;
      for (int ii = 0; ii < numComponents; ii++) {
         tmp = stats.at<int>(ii,CC_STAT_AREA);
         if (tmp > third) {
            if (tmp > second) {
               if (tmp > first) {
                  third = second;
                  ii3 = ii2;
                  second = first;
                  ii2 = ii1;
                  first = tmp;
                  ii1 = ii;
               } else {
                  third = second;
                  ii3 = ii2;
                  second = tmp;
                  ii2 = ii;
               }
            } else {
               third = tmp;
               ii3 = ii;
            }    
         }
      }    
      // Get the bounding box 
      int x = stats.at<int>(ii2,CC_STAT_LEFT);
      int y = stats.at<int>(ii2,CC_STAT_TOP);
      int w = stats.at<int>(ii2,CC_STAT_WIDTH);
      int h = stats.at<int>(ii2,CC_STAT_HEIGHT);
      /*
      x = x - w/2;
      w = 2*w;
      y = y - h/2;
      h = 2*h;*/
      //cout << "Width " << w << "\n";
      ball_frame = frame2(Rect(x,y,w,h));
      //imshow("test",ball_frame);
      //imshow("thresh",ball_frame>0);
      //waitKey(0);
      // Get approximate radius and center
      //int area = sum(ball_frame>0)[0];
      r = sqrt(second/3.1415926);
      //float cx = 0.25*(h+w);
      float cx = (float)centroids.at<double>(ii2,0)-x;
      float cy = (float)centroids.at<double>(ii2,1)-y;
      cent = Point(cx,cy);
   }
}

void process_data(
   const char * path_to_frames,
   //const char * prefix,
   int start_indx,
   vector<Mat>& im_vec,
   vector<float>& r_vec,
   vector<Point>& cent_vec) {
   
   //string root = string("frames/video");
   //root += to_string(vidIndx);
   //root += "_120fps_frame";
   string root = string(path_to_frames);
   root += "/";
   //root += prefix;
   string ext = string(".png");
   /*
   string root = string("Example");
   root += to_string(vidIndx);
   root += "/fullFrame_";
   string ext = string(".png");
   */

   try {
      int i = start_indx;
      Mat im_prev;
      //cout << root+to_string(i)+ext << "\n";
      Mat im = imread(root+to_string(i)+ext,0);
      i = start_indx+1;
      while (true) {
         im.copyTo(im_prev);
         im = imread(root+to_string(i)+ext,0);
         if (im.cols == 0) break;
         Mat ball_frame;
         float r=0;
         Point cent;
         detect_ball(im_prev,im,ball_frame,r,cent);
         //cout << "Frame " << i << "r=" << r << "cent=" << cent << "\n";
         im_vec.push_back(ball_frame);
         r_vec.push_back(r);
         cent_vec.push_back(cent);
         i++;
      }
   } catch (...) { cout << "Cannot load frames!!! Check if they exist\n"; }
}


void process_data(
   int vidIndx,
   vector<Mat>& im_vec,
   vector<float>& r_vec,
   vector<Point>& cent_vec) {
   
   //string root = string("frames/video");
   //root += to_string(vidIndx);
   //root += "_120fps_frame";
   string root = string("Example");
   root += to_string(vidIndx);
   root += "/fullFrame_";
   string ext = string(".png");

   try {
      int i = 0;
      Mat im_prev;
      Mat im = imread(root+to_string(i)+ext,0);
      i = 1;
      while (true) {
         im.copyTo(im_prev);
         im = imread(root+to_string(i)+ext,0);
         if (im.cols == 0) break;
         Mat ball_frame;
         float r=0;
         Point cent;
         detect_ball(im_prev,im,ball_frame,r,cent);
         //cout << "Frame " << i << "r=" << r << "cent=" << cent << "\n";
         im_vec.push_back(ball_frame);
         r_vec.push_back(r);
         cent_vec.push_back(cent);
         i++;
      }
   } catch (...) { cout << "Cannot load frames!!! Check if they exist\n"; }
}

/*
int main() {
   vector<Mat> im_vec;
   vector<float> r_vec;
   vector<Point> cent_vec;
   process_data(3,im_vec,r_vec,cent_vec);
}*/

File: python/detect_baseball.py
import numpy as np
import cv2

def detect_ball(frame1,frame2):
   '''
   Detect baseball with frame differencing. NOTE: We are assuming black 
   background with no net
   
   INPUT: consecutive frames (frame1 and frame2)
   OUTPUT: ball_frame, radius, center
   '''

   diff = frame2-frame1
   numBallPix = np.sum(diff > 100)
   if numBallPix > 150:
      # get the background
      bg = (diff < 30).astype(np.uint8)
      strel = np.zeros((25,25),dtype=np.uint8)
      bg_erode = cv2.erode(bg,strel,iterations=1)
      frame2_tmp = frame2*(bg_erode == 0)

      # Get the stats on binary image of the ball
      # This includes bounding box, radius, etc...
      ballonly = (frame2_tmp > 0).astype(np.uint8)
      output = cv2.connectedComponentsWithStats(ballonly,8,cv2.CV_32S)
      numComponents = output[0]
      labels = output[1]
      stats = output[2]
      centroid = output[3]

      # Get the second largest
      sort_indx = np.argsort(stats[:,cv2.CC_STAT_AREA])
      ball_indx = sort_indx[-2]
      
      # get bounding box 
      x = stats[ball_indx,cv2.CC_STAT_LEFT]
      y = stats[ball_indx,cv2.CC_STAT_TOP]
      w = stats[ball_indx,cv2.CC_STAT_WIDTH]
      h = stats[ball_indx,cv2.CC_STAT_HEIGHT]

      ball_frame = frame2[y:y+h,x:x+w]
      r = 1.*(stats[ball_indx,cv2.CC_STAT_AREA]/3.1415926)**0.5
      cent = centroid[ball_indx,:]
      cent[0] -= x
      cent[1] -= y

      return ball_frame,r,cent

def process_data(path_to_frames,prefix,start_indx):
   '''
   Get a list of the ball images, radii, and centers
   '''
   im_vec = []
   r_vec = []
   cent_vec = []
   try:
      im = cv2.imread(path_to_frames+'/{0}{1}.png'.format(prefix,start_indx),0)
      i = start_indx+1
      while True:
         im_prev = im.copy()
         im = cv2.imread(path_to_frames+'/{0}{1}.png'.format(prefix,i),0)
         
         if len(im[0,:]) == 0:
            break
         ball_frame,r,cent = detect_ball(im_prev,im)
         im_vec.append(ball_frame)
         r_vec.append(r)
         cent_vec.append(cent)
         i += 1
   except:
      pass 
   return im_vec,r_vec,cent_vec 

if __name__ == '__main__':
   path_to_frames = '/Users/james/Desktop/Consulting/repo/baseball_spin/SpinTests/002'
   prefix = ''
   im_vec,r_vec,cent_vec = process_data(path_to_frames,prefix,8)
   import matplotlib.pyplot as plt
   for ii in range(len(im_vec)):
      plt.figure()
      plt.imshow(im_vec[ii])
      print(r_vec[ii])
      print(cent_vec[ii])
      plt.show()
